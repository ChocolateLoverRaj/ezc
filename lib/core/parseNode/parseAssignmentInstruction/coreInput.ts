import coreParseCallAssignableInput from '../parseCallAssignable/coreInput'
import parseCallAssignable from '../parseCallAssignable/parseCallAssignable'
import Input from './Input'

const coreInput: Input = [parseCallAssignable(coreParseCallAssignableInput)]

export default coreInput
