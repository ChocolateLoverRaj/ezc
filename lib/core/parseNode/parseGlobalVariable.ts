import CoreTokenDatas from '../CoreTokenDatas'
import CoreTokenType from '../CoreTokenType'
import KeyWord from '../KeyWord'
import skip from '../splitAsyncIterator/skip'
import splitAsyncIterator from '../splitAsyncIterator/splitAsyncIterator'
import ConstantOrGlobal from './ConstantOrGlobal'
import CoreNodesWithData from './CoreNodesWithData'
import CoreNodeType from './CoreNodeType'
import coreTypeParsers from './coreTypeParsers'
import coreValueParsers from './coreValueParsers'
import Linkage from './Linkage'
import parseIdentifier from './parseIdentifier'
import tryNodeParsers from './tryNodeParsers'
import TryParseNode from './TryParseNode'

const parseGlobalVariable: TryParseNode<CoreNodesWithData[CoreNodeType.GLOBAL_VARIABLE]> =
async stream => {
  const asyncIterator = stream[Symbol.asyncIterator]()
  const splittedAsyncIterator = splitAsyncIterator(asyncIterator)
  const identifier = await parseIdentifier(splittedAsyncIterator.asyncIterable)
  if (identifier === undefined) return
  skip(splittedAsyncIterator, identifier.length)

  {
    const { done, value } = await splittedAsyncIterator.asyncIterable[Symbol.asyncIterator]().next()
    if (done === true) return
    if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
    const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
    if (!(data.enum === KeyWord && data.id === KeyWord.EQUALS)) return
  }
  skip(splittedAsyncIterator, 1)

  const allowedKeyWords = [
    KeyWord.PRIVATE,
    KeyWord.CONSTANT,
    KeyWord.GLOBAL,
    KeyWord.UNNAMED_ADDR
  ]
  const keyWords: KeyWord[] = []
  for await (const value of splittedAsyncIterator.asyncIterable) {
    if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
      const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
      if (!(data.enum === KeyWord && allowedKeyWords.includes(data.id))) return
      keyWords.push(data.id)
    } else {
      break
    }
  }
  skip(splittedAsyncIterator, keyWords.length)
  const isConstant = keyWords.includes(KeyWord.CONSTANT)
  const isGlobal = keyWords.includes(KeyWord.GLOBAL)
  // Must have exactly one of constant or global
  if (Number(isConstant) + Number(isGlobal) !== 1) {
    return
  }

  const type = await tryNodeParsers(
    coreTypeParsers)(splittedAsyncIterator.asyncIterable)
  if (type === undefined) return
  skip(splittedAsyncIterator, type.length)

  const value = await tryNodeParsers(coreValueParsers)(splittedAsyncIterator.asyncIterable)
  if (value === undefined) return
  skip(splittedAsyncIterator, value.length)

  return {
    node: {
      type: {
        enum: CoreNodeType,
        id: CoreNodeType.GLOBAL_VARIABLE
      },
      data: {
        // FIXME: No align logic
        align: undefined,
        constantOrGlobal: isConstant
          ? ConstantOrGlobal.CONSTANT
          : ConstantOrGlobal.GLOBAL,
        identifier: identifier.node,
        // FIXME: No other linkage types
        linkage: Linkage.PRIVATE,
        type: type.node,
        unnamed_addr: keyWords.includes(KeyWord.UNNAMED_ADDR),
        value: value.node
      }
    },
    length: identifier.length + 1 + keyWords.length + type.length + value.length
  }
}

export default parseGlobalVariable
