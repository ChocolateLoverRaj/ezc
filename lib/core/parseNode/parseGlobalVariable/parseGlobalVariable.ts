import CoreTokenDatas from '../../CoreTokenDatas'
import CoreTokenType from '../../CoreTokenType'
import CoreKeyWord from '../../CoreKeyWord'
import skip from '../../splitAsyncIterator/skip'
import splitAsyncIterator from '../../splitAsyncIterator/splitAsyncIterator'
import ConstantOrGlobal from '../ConstantOrGlobal'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import Linkage from '../Linkage'
import parseIdentifier from '../parseIdentifier'
import tryNodeParsers from '../tryNodeParsers'
import TryParseNode from '../TryParseNode'
import Input from './Input'

const parseGlobalVariable = (
  { typeParsers, valueParsers }: Input
): TryParseNode<CoreNodesWithData[CoreNodeType.GLOBAL_VARIABLE]> =>
  async stream => {
    const asyncIterator = stream[Symbol.asyncIterator]()
    const splittedAsyncIterator = splitAsyncIterator(asyncIterator)
    const identifier = await parseIdentifier(splittedAsyncIterator.asyncIterable)
    if (identifier === undefined) return
    skip(splittedAsyncIterator, identifier.length)

    {
      const { done, value } =
        await splittedAsyncIterator.asyncIterable[Symbol.asyncIterator]().next()
      if (done === true) return
      if (!(value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD)) return
      const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
      if (!(data.enum === CoreKeyWord && data.id === CoreKeyWord.EQUALS)) return
    }
    skip(splittedAsyncIterator, 1)

    const allowedKeyWords = [
      CoreKeyWord.PRIVATE,
      CoreKeyWord.CONSTANT,
      CoreKeyWord.GLOBAL,
      CoreKeyWord.UNNAMED_ADDR
    ]
    const keyWords: CoreKeyWord[] = []
    for await (const value of splittedAsyncIterator.asyncIterable) {
      if (value.type.enum === CoreTokenType && value.type.id === CoreTokenType.KEY_WORD) {
        const data = value.data as CoreTokenDatas[CoreTokenType.KEY_WORD]
        if (!(data.enum === CoreKeyWord && allowedKeyWords.includes(data.id))) return
        keyWords.push(data.id)
      } else {
        break
      }
    }
    skip(splittedAsyncIterator, keyWords.length)
    const isConstant = keyWords.includes(CoreKeyWord.CONSTANT)
    const isGlobal = keyWords.includes(CoreKeyWord.GLOBAL)
    // Must have exactly one of constant or global
    if (Number(isConstant) + Number(isGlobal) !== 1) {
      return
    }

    const type = await tryNodeParsers(typeParsers)(splittedAsyncIterator.asyncIterable)
    if (type === undefined) return
    skip(splittedAsyncIterator, type.length)

    const value = await tryNodeParsers(valueParsers)(splittedAsyncIterator.asyncIterable)
    if (value === undefined) return
    skip(splittedAsyncIterator, value.length)

    return {
      node: {
        type: {
          enum: CoreNodeType,
          id: CoreNodeType.GLOBAL_VARIABLE
        },
        data: {
        // FIXME: No align logic
          align: undefined,
          constantOrGlobal: isConstant
            ? ConstantOrGlobal.CONSTANT
            : ConstantOrGlobal.GLOBAL,
          identifier: identifier.node,
          // FIXME: No other linkage types
          linkage: Linkage.PRIVATE,
          type: type.node,
          unnamed_addr: keyWords.includes(CoreKeyWord.UNNAMED_ADDR),
          value: value.node
        }
      },
      length: identifier.length + 1 + keyWords.length + type.length + value.length
    }
  }

export default parseGlobalVariable
