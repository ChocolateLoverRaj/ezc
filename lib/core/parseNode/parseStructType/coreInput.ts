import coreTypeParsers from '../coreTypeParsers'
import Input from './Input'

const coreInput: Input = coreTypeParsers

export default coreInput
