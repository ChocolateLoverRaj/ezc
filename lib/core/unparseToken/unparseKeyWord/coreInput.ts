import CoreKeyWord from '../../CoreKeyWord'
import Input from './Input'

const coreKeyWordToStringMap: Map<CoreKeyWord, string> = new Map([
  [CoreKeyWord.ALIGN, 'align'],
  [CoreKeyWord.CALL, 'call'],
  [CoreKeyWord.COMMA, ','],
  [CoreKeyWord.CONSTANT, 'constant'],
  [CoreKeyWord.DECLARE, 'declare'],
  [CoreKeyWord.DEFINE, 'define'],
  [CoreKeyWord.EQUALS, '='],
  [CoreKeyWord.GET_ELEMENT_PTR, 'getelementptr'],
  [CoreKeyWord.GLOBAL, 'global'],
  [CoreKeyWord.INBOUNDS, 'inbounds'],
  [CoreKeyWord.NO_ALIAS, 'noalias'],
  [CoreKeyWord.NO_CAPTURE, 'nocapture'],
  [CoreKeyWord.NO_FREE, 'nofree'],
  [CoreKeyWord.PRIVATE, 'private'],
  [CoreKeyWord.PTR, 'ptr'],
  [CoreKeyWord.RETURN, 'ret'],
  [CoreKeyWord.UNNAMED_ADDR, 'unnamed_addr'],
  [CoreKeyWord.VOID, 'void'],
  [CoreKeyWord.X, 'x'],
  [CoreKeyWord.OPEN_PARENTHESIS, '('],
  [CoreKeyWord.CLOSE_PARENTHESIS, ')'],
  [CoreKeyWord.OPEN_BRACKET, '['],
  [CoreKeyWord.CLOSE_BRACKET, ']'],
  [CoreKeyWord.OPEN_CURLY_BRACKET, '{'],
  [CoreKeyWord.CLOSE_CURLY_BRACKET, '}']
])

const coreInput: Input = new Map([
  [CoreKeyWord, coreKeyWordToStringMap]
])

export default coreInput
