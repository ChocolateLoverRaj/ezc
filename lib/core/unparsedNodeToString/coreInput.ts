import coreUnparseTokenInput from '../unparseToken/unparseToken/coreInput'
import coreNodeUnparsers from './coreNodeUnparsers'
import Input from './Input'

const coreInput: Input = {
  unparseTokenInput: coreUnparseTokenInput,
  nodeUnparsers: coreNodeUnparsers
}

export default coreInput
