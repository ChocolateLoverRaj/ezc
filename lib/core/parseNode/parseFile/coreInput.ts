import coreParseDeclareInput from '../parseDeclare/coreInput'
import parseDeclare from '../parseDeclare/parseDeclare'
import parseFunction from '../parseFunction/parseFunction'
import coreParseGlobalVariableInput from '../parseGlobalVariable/coreInput'
import parseGlobalVariable from '../parseGlobalVariable/parseGlobalVariable'
import tryNodeParsers from '../tryNodeParsers'
import Input from './Input'
import coreParseFunctionInput from '../parseFunction/coreInput'

const coreInput: Input = tryNodeParsers([
  parseDeclare(coreParseDeclareInput),
  parseGlobalVariable(coreParseGlobalVariableInput),
  parseFunction(coreParseFunctionInput)
])

export default coreInput
