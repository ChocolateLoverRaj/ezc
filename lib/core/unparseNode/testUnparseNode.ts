import TryParseNode from '../parseNode/TryParseNode'
import EnumItemWithData from '../EnumItemWithData'
import UnparsedPartType from './UnparsedPartType'
import unparsedNodeToString from '../unparsedNodeToString/unparsedNodeToString'
import coreInput from '../unparsedNodeToString/coreInput'
import parseAllTokens from '../tryGetToken/parseAllTokens'
import coreTryers from '../tryGetToken/coreTryers'
import arrayToAsyncIterable from '../arrayToAsyncIterable'

const testUnparseNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  expect(unparsedNodeToString(coreInput)([{
    type: UnparsedPartType.NODE,
    data: (await parseNode(
      parseAllTokens(coreTryers)(
        arrayToAsyncIterable([code])) as any))?.node as EnumItemWithData
  }])).toMatchSnapshot()
}

export default testUnparseNode
