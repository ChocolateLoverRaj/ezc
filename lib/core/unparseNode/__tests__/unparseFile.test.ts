import coreInput from '../../parseNode/parseGlobalVariable/coreInput'
import parseGlobalVariable from '../../parseNode/parseGlobalVariable/parseGlobalVariable'
import testUnparseNode from '../testUnparseNode'

test('just global variable', async () => {
  await testUnparseNode(parseGlobalVariable(coreInput), '@0 = private constant unnamed_addr i32 69')
})
