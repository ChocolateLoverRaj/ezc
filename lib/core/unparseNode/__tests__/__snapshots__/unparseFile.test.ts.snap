// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`just global variable 1`] = `
Array [
  Object {
    "data": Object {
      "align": undefined,
      "constantOrGlobal": 1,
      "identifier": Object {
        "data": Object {
          "name": "0",
          "type": 0,
        },
        "type": Object {
          "enum": Object {
            "0": "FILE",
            "1": "DECLARE",
            "10": "IDENTIFIER",
            "11": "NUMBER",
            "12": "VOID_TYPE",
            "13": "BLOCK",
            "14": "RETURN_INSTRUCTION",
            "15": "CALL_ASSIGNABLE",
            "16": "ASSIGNMENT_INSTRUCTION",
            "17": "INPUT",
            "18": "INPUT_FLAG",
            "2": "FUNCTION",
            "3": "INTEGER_TYPE",
            "4": "GLOBAL_VARIABLE",
            "5": "STRING",
            "6": "FLOAT_TYPE",
            "7": "POINTER_TYPE",
            "8": "ARRAY_TYPE",
            "9": "STRUCT_TYPE",
            "ARRAY_TYPE": 8,
            "ASSIGNMENT_INSTRUCTION": 16,
            "BLOCK": 13,
            "CALL_ASSIGNABLE": 15,
            "DECLARE": 1,
            "FILE": 0,
            "FLOAT_TYPE": 6,
            "FUNCTION": 2,
            "GLOBAL_VARIABLE": 4,
            "IDENTIFIER": 10,
            "INPUT": 17,
            "INPUT_FLAG": 18,
            "INTEGER_TYPE": 3,
            "NUMBER": 11,
            "POINTER_TYPE": 7,
            "RETURN_INSTRUCTION": 14,
            "STRING": 5,
            "STRUCT_TYPE": 9,
            "VOID_TYPE": 12,
          },
          "id": 10,
        },
      },
      "linkage": 0,
      "type": Object {
        "data": 32,
        "type": Object {
          "enum": Object {
            "0": "FILE",
            "1": "DECLARE",
            "10": "IDENTIFIER",
            "11": "NUMBER",
            "12": "VOID_TYPE",
            "13": "BLOCK",
            "14": "RETURN_INSTRUCTION",
            "15": "CALL_ASSIGNABLE",
            "16": "ASSIGNMENT_INSTRUCTION",
            "17": "INPUT",
            "18": "INPUT_FLAG",
            "2": "FUNCTION",
            "3": "INTEGER_TYPE",
            "4": "GLOBAL_VARIABLE",
            "5": "STRING",
            "6": "FLOAT_TYPE",
            "7": "POINTER_TYPE",
            "8": "ARRAY_TYPE",
            "9": "STRUCT_TYPE",
            "ARRAY_TYPE": 8,
            "ASSIGNMENT_INSTRUCTION": 16,
            "BLOCK": 13,
            "CALL_ASSIGNABLE": 15,
            "DECLARE": 1,
            "FILE": 0,
            "FLOAT_TYPE": 6,
            "FUNCTION": 2,
            "GLOBAL_VARIABLE": 4,
            "IDENTIFIER": 10,
            "INPUT": 17,
            "INPUT_FLAG": 18,
            "INTEGER_TYPE": 3,
            "NUMBER": 11,
            "POINTER_TYPE": 7,
            "RETURN_INSTRUCTION": 14,
            "STRING": 5,
            "STRUCT_TYPE": 9,
            "VOID_TYPE": 12,
          },
          "id": 3,
        },
      },
      "unnamed_addr": true,
      "value": Object {
        "data": 69,
        "type": Object {
          "enum": Object {
            "0": "FILE",
            "1": "DECLARE",
            "10": "IDENTIFIER",
            "11": "NUMBER",
            "12": "VOID_TYPE",
            "13": "BLOCK",
            "14": "RETURN_INSTRUCTION",
            "15": "CALL_ASSIGNABLE",
            "16": "ASSIGNMENT_INSTRUCTION",
            "17": "INPUT",
            "18": "INPUT_FLAG",
            "2": "FUNCTION",
            "3": "INTEGER_TYPE",
            "4": "GLOBAL_VARIABLE",
            "5": "STRING",
            "6": "FLOAT_TYPE",
            "7": "POINTER_TYPE",
            "8": "ARRAY_TYPE",
            "9": "STRUCT_TYPE",
            "ARRAY_TYPE": 8,
            "ASSIGNMENT_INSTRUCTION": 16,
            "BLOCK": 13,
            "CALL_ASSIGNABLE": 15,
            "DECLARE": 1,
            "FILE": 0,
            "FLOAT_TYPE": 6,
            "FUNCTION": 2,
            "GLOBAL_VARIABLE": 4,
            "IDENTIFIER": 10,
            "INPUT": 17,
            "INPUT_FLAG": 18,
            "INTEGER_TYPE": 3,
            "NUMBER": 11,
            "POINTER_TYPE": 7,
            "RETURN_INSTRUCTION": 14,
            "STRING": 5,
            "STRUCT_TYPE": 9,
            "VOID_TYPE": 12,
          },
          "id": 11,
        },
      },
    },
    "type": Object {
      "enum": Object {
        "0": "FILE",
        "1": "DECLARE",
        "10": "IDENTIFIER",
        "11": "NUMBER",
        "12": "VOID_TYPE",
        "13": "BLOCK",
        "14": "RETURN_INSTRUCTION",
        "15": "CALL_ASSIGNABLE",
        "16": "ASSIGNMENT_INSTRUCTION",
        "17": "INPUT",
        "18": "INPUT_FLAG",
        "2": "FUNCTION",
        "3": "INTEGER_TYPE",
        "4": "GLOBAL_VARIABLE",
        "5": "STRING",
        "6": "FLOAT_TYPE",
        "7": "POINTER_TYPE",
        "8": "ARRAY_TYPE",
        "9": "STRUCT_TYPE",
        "ARRAY_TYPE": 8,
        "ASSIGNMENT_INSTRUCTION": 16,
        "BLOCK": 13,
        "CALL_ASSIGNABLE": 15,
        "DECLARE": 1,
        "FILE": 0,
        "FLOAT_TYPE": 6,
        "FUNCTION": 2,
        "GLOBAL_VARIABLE": 4,
        "IDENTIFIER": 10,
        "INPUT": 17,
        "INPUT_FLAG": 18,
        "INTEGER_TYPE": 3,
        "NUMBER": 11,
        "POINTER_TYPE": 7,
        "RETURN_INSTRUCTION": 14,
        "STRING": 5,
        "STRUCT_TYPE": 9,
        "VOID_TYPE": 12,
      },
      "id": 4,
    },
  },
]
`;
